kernel/aster-nix/src/vm/vmo/mod.rs

```diff
@@ -127,219 +135,370 @@ bitflags! {
+    ) -> Result<Arc<Mutex<XArray<VmFrame, VmoMark>>>> {
+        let child_vmo_start = range.start;
+        let child_vmo_end = range.end;
+        debug_assert!(child_vmo_start % PAGE_SIZE == 0);
+        debug_assert!(child_vmo_end % PAGE_SIZE == 0);
+        if child_vmo_start % PAGE_SIZE != 0 || child_vmo_end % PAGE_SIZE != 0 {
+            return_errno_with_message!(Errno::EINVAL, "vmo range does not aligned with PAGE_SIZE");
+        }
+
+        match child_type {
+            ChildType::Slice => {
+                let parent_vmo_size = self.size();
+                // A slice child should be inside parent vmo's range
+                debug_assert!(child_vmo_end <= parent_vmo_size);
+                if child_vmo_end > parent_vmo_size {
+                    return_errno_with_message!(
+                        Errno::EINVAL,
+                        "slice child vmo cannot exceed parent vmo's size"
+                    );
+                }
+                // Condition 1.
+                Ok(self.pages.clone())
+            }
+            ChildType::Cow => {
+                let mut pages = self.pages.lock();
+                let parent_vmo_size = self.size();
+                // A copy on Write child should intersect with parent vmo
+                debug_assert!(child_vmo_start <= parent_vmo_size);
+                if child_vmo_start > parent_vmo_size {
+                    return_errno_with_message!(
+                        Errno::EINVAL,
+                        "COW vmo should overlap with its parent"
+                    );
+                }
+                let self_is_cow = pages.is_marked(VmoMark::CowVmo);
+
+                if self_is_cow {
+                    // Condition 2.
+                    pages.unset_mark_all(VmoMark::ExclusivePage);
+                    return Ok(Arc::new(Mutex::new(pages.clone())));
+                }
+
+                if self.pager.is_some() {
+                    // Condition 3.
+                    let mut cloned_pages = pages.clone();
+                    cloned_pages.set_mark(VmoMark::CowVmo);
+                    return Ok(Arc::new(Mutex::new(cloned_pages)));
+                }
+
+                // Condition 4.
+                pages.set_mark(VmoMark::CowVmo);
+                Ok(Arc::new(Mutex::new(pages.clone())))
+            }
+        }
+    }
+
+    /// Resize current Vmo to target size.
     pub fn resize(&self, new_size: usize) -> Result<()> {
         assert!(self.flags.contains(VmoFlags::RESIZABLE));
         let new_size = new_size.align_up(PAGE_SIZE);
+
+        let pages = self.pages.lock();
         let old_size = self.size();
         if new_size == old_size {
             return Ok(());
         }
-
         if new_size < old_size {
-            self.decommit(new_size..old_size)?;
-            self.inner.lock().size = new_size;
-        } else {
-            self.inner.lock().size = new_size;
+            self.decommit_pages(pages, new_size..old_size)?;
```

Comment:
> To me, this logic also gives me the feeling that VMO should never have a mutable size that can be changed arbitrarily.

Thank you for your comments. Actually, I have the same concern with you. More specifically, my opinion is that a vmo that can be mapped cannot resize. As I mentioned above:
> Mapping a resizable vmo may easily introduce bugs (cause a bug to this pr) and make it difficult to define whether mapping is an anonymous mapping or a filebacked mapping.

So the logics of `decommit` here should work for those vmo can not be mapped, and we also plan to remove the VMO usage in VmMapping. This is a big project and will be done gradually in the future PR.

Currently `ChildType::Slice` is designed for shared mapping and is used in #617 . But I think these child vmo semantics will be removed in the future PR.

Reply:
Thanks for the explanation.

> So the logics of decommit here should work for those vmo can not be mapped

Not sure about what it means. Could you please give a use case about VMOs that cannot be mapped?

I was once thinking once resizing VMO is necessary for the `brk()` system call that adjusts the heap size. However, the heap is certainly mapped.

> Currently ChildType::Slice is designed for shared mapping and is used in https://github.com/asterinas/asterinas/pull/617 .

I would suggest to isolate this special case in the type definition (instead of solely using flags, which is obscure and error-prune).

I am not very clear about the future of VMO, but take the current design in this PR as an example:
```rust
struct Vmo_ {
    pages: Pages,
    ...
}

enum Pages {
    // Shared pages across multiple (non-resizable) VMOs.
    Shared(Arc<Mutex<XArray<VmFrame, VmoMark>>>),
    // Private pages and the size of the current (resizable) VMO.
    Private(Mutex<(XArray<VmFrame, VmoMark>, usize)>),
}

impl Pages {
    fn with<R>(&self, f: FnOnce(&mut XArray<VmFrame, VmoMark>) -> R) -> R {
        ...
    }
}
```

This does introduce some complexities in the type definition. But I think it makes things clearer and resolves the original concern.

> Given `pages: Arc<_>` can be shared by multiple VMOs (`ChildType::Slice`), why should we decommit pages just because the size of one VMO shrinks? After all, other VMOs sharing the same `pages` may still want to use these pages.

Reply:
> Could you please give a use case about VMOs that cannot be mapped?

Based on our current understanding, a resizable VMO should not be mapped. For example, let's say we have a VMO which is initialized based on part of a file. If the VMO is resized and becomes larger, the additional portion will turn into anonymous pages. At that point, mapping the VMO becomes problematic for two reasons. Firstly, it's challenging to distinguish whether the mapping is a file mapping or an anonymous mapping, which can lead to semantic confusion. Secondly, additional operations are required during VMO read and write actions to determine whether the page originated from the file or was generated by the resize.

Reply:
Thanks for your detailed explanation.

As you said, `VmoChildOptions` documents that the slice VMO child and _its parent_ cannot be resizable. However, I do not believe that this requirement is ever checked when the slice VMO is created by `Vmo::new_slice_child` or `VmoChildOptions::new_slice_rights`. (At least I cannot easily find such checks.)

So there is nothing to prevent one from creating a slice VMO from a _resizable parent_ by calling the above public APIs. That's why we should encode the requirement in type definitions, which forces the requirement to be checked whenever necessary.

> I wonder if the adjustments to this logic could be made in a later PR?

My main concern is that, it is this PR that introduces the following two fields in `Vmo_`:
```rust
pub(super) struct Vmo_ {
    ...
    /// The contiguous pages where the VMO resides. Pages are managed by XArray,
    /// and can be shared among slice children.
    pages: Arc<Mutex<XArray<VmFrame, VmoMark>>>,
    /// The size of current `Vmo_`.
    size: AtomicUsize,
}
```
where:
 1. `size` (outside the `Arc<_>`) needs to be consistent with the `XArray` (inside the `Arc<_>`), as things that `resize()` wants to do;
 2. `size` (outside the `Arc<_>`) needs to be protected by the `Mutex` (inside the `Arc<_>`), to avoid race conditions.

To me, this is very wired and error-prone logic, so I would suggest addressing the problem as it is actually (or at least appears to be) introduced by this PR.

But I also understand that the associated VM logic is in the process of being redesigned. As you said, after the redesign we may not even need VMO slicing, so the problem will be solved automatically without any effort. So if you prefer to leave the solution in a future PR, maybe we can at least add some assertions and comments in this PR.

For example, check that the VMO is not resizable before cloning `pages: Arc<_>':
```rust
        match child_type {
            ChildType::Slice => {
                ...
                // FIXME: Check the VMO is not resizable

                // Condition 1.
                Ok(self.pages.clone())
```

Add comments for `size: AtomicUsize`:
```rust
pub(super) struct Vmo_ {
    ...
    /// The contiguous pages where the VMO resides. Pages are managed by XArray,
    /// and can be shared among slice children.
    pages: Arc<Mutex<XArray<VmFrame, VmoMark>>>,
    /// The size of current `Vmo_`.
    ///
    /// FIXME: Explain
    ///   1. `size` is protected by the lock in `pages`;
    ///   2. VMO is not resizable, if pages are shared across multiple VMOs
    ///      (in other words, if the VMO is a slice parent or child).
    size: AtomicUsize,
}
```

(But I feel that if you have time to write such detailed comments, you can spend similar time just fixing the type definition. I have skimmed the implementation and feel that it is not very hard to do..?)

Or you can try to ask for comments from other reviewers. This is not really a big problem, because the VMO redesign is coming. I will not insist on it if others find the current design somehow acceptable.